src/cli_util.rs

```diff
@@ -358,6 +358,54 @@ impl Debug for ChromeTracingFlushGuard {
     }
 }
 
+// The default field formatter prints all fields. The `#[instrument]` macro
+// includes each parameter's name and value as fields, which is too noisy to
+// read. Note that, as written, the list of function names in the stack trace
+// (separated by `:`) is still printed.
```

Comment:
I think traces in `git.rs` are supposed to include parameters. For example,
```
2023-07-23T21:42:36.645567Z  INFO cmd_git_fetch{args=GitFetchArgs { branch: [], remotes: [] }}:fetch{remote_name="upstream" branch_name_globs=None git_settings=GitSettings { auto_local_branch: false }}: jj_lib::git: using ssh_key_from_agent username="git"
```
becomes
```
2023-07-23T21:42:04.388399Z  INFO run_command:cmd_git_fetch:fetch: jj_lib::git: using ssh_key_from_agent
```

The latter is rather useless.

Reply:
@yuja you're right, that's unfortunate. But immediately before this commit, it's also useless (unless you can read all this 😄):

```
2023-07-23T22:19:23.977782Z DEBUG run:run_internal{ui=Ui { color: true, pager_cmd: Structured { env: {"LESSCHARSET": "utf-8"}, command: NonEmptyCommandArgsVec(["less", "-FRX"]) }, paginate: Auto, progress_indicator: true, formatter_factory: FormatterFactory { kind: Color { rules: [(["diff", "file_header"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["working_copies"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["git_head"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["git_refs"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["branches"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["working_copy", "divergent"], Style { fg_color: Some(Red), bg_color: None, bold: None, underlined: None }), (["tags"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["working_copy", "branch"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["working_copy", "working_copies"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["conflict"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["diff", "added"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["working_copy", "tags"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["working_copy"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["timestamp"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["op_log", "time"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["email"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["conflict_description", "difficult"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["op_log", "user"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["branch"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["working_copy", "conflict"], Style { fg_color: Some(Red), bg_color: None, bold: None, underlined: None }), (["prefix"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["op_log", "current_operation", "id"], Style { fg_color: Some(Blue), bg_color: None, bold: None, underlined: None }), (["rest"], Style { fg_color: Some(DarkGrey), bg_color: None, bold: None, underlined: None }), (["divergent", "change_id"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["warning"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["working_copy", "git_refs"], Style { fg_color: Some(Green), bg_color: None, bold: None, underlined: None }), (["username"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["op_log", "id"], Style { fg_color: Some(DarkBlue), bg_color: None, bold: None, underlined: None }), (["working_copy", "empty"], Style { fg_color: Some(Green), bg_color: None, bold: None, underlined: None }), (["working_copy", "commit_id"], Style { fg_color: Some(Blue), bg_color: None, bold: None, underlined: None }), (["error"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["divergent", "rest"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["diff", "removed"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["working_copy", "branches"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["op_log", "current_operation"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["empty"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["working_copy", "timestamp"], Style { fg_color: Some(Cyan), bg_color: None, bold: None, underlined: None }), (["op_log", "current_operation", "time"], Style { fg_color: Some(Cyan), bg_color: None, bold: None, underlined: None }), (["hidden", "prefix"], Style { fg_color: Some(Reset), bg_color: None, bold: None, underlined: None }), (["change_id"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["hint"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["divergent", "hidden", "rest"], Style { fg_color: Some(DarkGrey), bg_color: None, bold: Some(false), underlined: None }), (["diff", "header"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["working_copy", "email"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["divergent", "prefix"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: Some(true) }), (["working_copy", "change_id"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["commit_id"], Style { fg_color: Some(DarkBlue), bg_color: None, bold: None, underlined: None }), (["conflict_description"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["op_log", "current_operation", "user"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["diff", "hunk_header"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["divergent", "hidden", "prefix"], Style { fg_color: Some(Reset), bg_color: None, bold: None, underlined: Some(false) }), (["working_copy", "divergent", "change_id"], Style { fg_color: Some(Red), bg_color: None, bold: None, underlined: None }), (["divergent"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["divergent", "hidden"], Style { fg_color: Some(Reset), bg_color: None, bold: Some(true), underlined: None }), (["diff", "modified"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None })] } }, output: Terminal { stdout: Stdout { .. }, stderr: Stderr { .. } } }}:run_command{ui=Ui { color: true, pager_cmd: Structured { env: {"LESSCHARSET": "utf-8"}, command: NonEmptyCommandArgsVec(["less", "-FRX"]) }, paginate: Auto, progress_indicator: true, formatter_factory: FormatterFactory { kind: Color { rules: [(["timestamp"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["working_copy", "tags"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["diff", "modified"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["prefix"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["email"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["divergent", "hidden"], Style { fg_color: Some(Reset), bg_color: None, bold: Some(true), underlined: None }), (["diff", "file_header"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["working_copy", "empty"], Style { fg_color: Some(Green), bg_color: None, bold: None, underlined: None }), (["divergent", "change_id"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["error"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["op_log", "current_operation", "id"], Style { fg_color: Some(Blue), bg_color: None, bold: None, underlined: None }), (["working_copy", "branches"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["branch"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["diff", "added"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["op_log", "user"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["working_copy", "divergent", "change_id"], Style { fg_color: Some(Red), bg_color: None, bold: None, underlined: None }), (["git_head"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["conflict_description"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["working_copy", "divergent"], Style { fg_color: Some(Red), bg_color: None, bold: None, underlined: None }), (["op_log", "current_operation", "time"], Style { fg_color: Some(Cyan), bg_color: None, bold: None, underlined: None }), (["tags"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["working_copy", "change_id"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["diff", "header"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["username"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["commit_id"], Style { fg_color: Some(DarkBlue), bg_color: None, bold: None, underlined: None }), (["empty"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["conflict_description", "difficult"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["git_refs"], Style { fg_color: Some(DarkGreen), bg_color: None, bold: None, underlined: None }), (["hint"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["working_copy"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["working_copy", "email"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["diff", "hunk_header"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["conflict"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["working_copy", "working_copies"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["divergent"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["rest"], Style { fg_color: Some(DarkGrey), bg_color: None, bold: None, underlined: None }), (["change_id"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["divergent", "hidden", "rest"], Style { fg_color: Some(DarkGrey), bg_color: None, bold: Some(false), underlined: None }), (["working_copy", "commit_id"], Style { fg_color: Some(Blue), bg_color: None, bold: None, underlined: None }), (["working_copy", "conflict"], Style { fg_color: Some(Red), bg_color: None, bold: None, underlined: None }), (["divergent", "rest"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["divergent", "hidden", "prefix"], Style { fg_color: Some(Reset), bg_color: None, bold: None, underlined: Some(false) }), (["op_log", "current_operation"], Style { fg_color: None, bg_color: None, bold: Some(true), underlined: None }), (["working_copy", "branch"], Style { fg_color: Some(Magenta), bg_color: None, bold: None, underlined: None }), (["op_log", "time"], Style { fg_color: Some(DarkCyan), bg_color: None, bold: None, underlined: None }), (["warning"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["diff", "removed"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: None }), (["working_copies"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["op_log", "id"], Style { fg_color: Some(DarkBlue), bg_color: None, bold: None, underlined: None }), (["working_copy", "timestamp"], Style { fg_color: Some(Cyan), bg_color: None, bold: None, underlined: None }), (["hidden", "prefix"], Style { fg_color: Some(Reset), bg_color: None, bold: None, underlined: None }), (["op_log", "current_operation", "user"], Style { fg_color: Some(DarkYellow), bg_color: None, bold: None, underlined: None }), (["working_copy", "git_refs"], Style { fg_color: Some(Green), bg_color: None, bold: None, underlined: None }), (["branches"], Style { fg_color: Some(DarkMagenta), bg_color: None, bold: None, underlined: None }), (["divergent", "prefix"], Style { fg_color: Some(DarkRed), bg_color: None, bold: None, underlined: Some(true) })] } }, output: Terminal { stdout: Stdout { .. }, stderr: Stderr { .. } } } command_helper=CommandHelper { .. }}:cmd_git_fetch{args=GitFetchArgs { branch: [], remotes: [] }}:fetch{remote_name="origin" branch_name_globs=None git_settings=GitSettings { auto_local_branch: false }}: jj_lib::git: import_refs
```

I tried to limit the amount of `Debug` information in the `#[instrument]` traces, but I think it's still too much, like in the commit message example:

```
    2023-07-23T10:23:05.639140Z  INFO run:run_internal{ui=Ui { .. }}:run_command{ui=Ui { .. } command_helper=CommandHelper { .. }}:cmd_status{ui=Ui { .. } command=CommandHelper { .. } _args=StatusArgs}:workspace_helper_internal{snapshot=true}:snapshot{self=WorkspaceCommandHelper { .. } ui=Ui { .. }}:snapshot_working_copy{self=WorkspaceCommandHelper { .. } ui=Ui { .. }}:snapshot{self=TreeState { working_copy_path: "/Users/waleed/Workspace/nixpkgs", state_path: "/Users/waleed/Workspace/jj/../nixpkgs/.jj/working_copy", tree_id: TreeId("52cbddc7d9c7d51b9ef612f4e8bfd69f2140ee0c"), own_mtime: MillisSinceEpoch(1690107765806), watchman_clock: Some(WatchmanClock { watchman_clock: Some(StringClock("c:1689499456:1591:1:4356")) }), .. } options=SnapshotOptions { fsmonitor_kind: Some(Watchman), .. }}:make_fsmonitor_matcher{self=T
reeState { working_copy_path: "/Users/waleed/Workspace/nixpkgs", state_path: "/Users/waleed/Workspace/jj/../nixpkgs/.jj/working_copy", tree_id: TreeId("52cbddc7d9c7d51b9ef612f4e8bfd69f2140ee0c"), own_mtime: MillisSinceEpoch(1690107765806), watchman_clock: Some(WatchmanClock { watchman_clock: Some(StringClock("c:1689499456:1591:1:4356")) }), .. } fsmonitor_kind=Some(Watchman)}:query_watchman{self=TreeState { working_copy_path: "/Users/waleed/Workspace/nixpkgs", state_path: "/Users/waleed/Workspace/jj/../nixpkgs/.jj/working_copy", tree_id: TreeId("52cbddc7d9c7d51b9ef612f4e8bfd69f2140ee0c"), own_mtime: MillisSinceEpoch(1690107765806), watchman_clock: Some(WatchmanClock { watchman_clock: Some(StringClock("c:1689499456:1591:1:4356")) }), .. }}:query_changed_files{self=watchman::Fsmonitor { .. } previous_clock=Some(Clock(Spec(StringClock("c:1689499456:1591:1:4356"))))}: jj_lib::fsmonitor::watchman: Querying Watchman for changed files...
```

I think the best way is to explicitly include fields in `debug` instead of relying on `#[instrument]` like this existing line:

```
            tracing::debug!(?globs, ?result, "globs as regex");
```

This kind of thing renders at the end of the line instead of with the spans, so it must be possible to render one but not the other, but I can't figure out how.

Reply:
I suppose the answer might just be to add more `skip_all`s until the logging is at a manageable size.

Reply:
Can't we skip uninteresting arguments to reduce noise level? Do we have parameters that are important for `JJ_TRACE`, but not for `jj -v`?

I'm not familiar with tracing API, so I don't know the best practice we can follow.

Reply:
I added a ton of `skip_all`s and I think the logging output is useful now.

Watchman example:

```
2023-07-25T14:59:06.675076Z  INFO run:run_internal:run_command:cmd_status:workspace_helper_internal{snapshot=true}:snapshot:snapshot_working_copy:snapshot:make_fsmonitor_matcher:query_watchman:query_changed_files{previous_clock=Some(Clock(Spec(StringClock("c:1689499456:1591:1:8450"))))}: jj_lib::fsmonitor::watchman: Querying Watchman for changed files...
```

`jj git fetch` example:

```
2023-07-25T14:58:47.369777Z DEBUG run:run_internal:run_command:cmd_git_fetch{args=GitFetchArgs { branch: [], remotes: [] }}:fetch{remote_name="origin" branch_name_globs=None git_settings=GitSettings { auto_local_branch: false }}: jj_lib::git: remote.download
```

They span 2-3 lines in an 80-char terminal, which I think is acceptable.

Reply:
> Do we have parameters that are important for JJ_TRACE

In principle, you can ask the Chrome tracing layer to record arguments, which you can then see in the trace. I've used this in the past to debug why an operation was especially slow for a certain commit, because I could find the slow operation in the trace and click it to see what the associated commit ID was. I think it's not worth prematurely preserving this behavior at the expense of ridiculously-long verbose logging. We could just add the extra logs during profiling later (assuming that we're not trying to debug a user's issue by capturing their trace for whom recompiling is not an option).

Reply:
> I added a ton of skip_alls and I think the logging output is useful now.

Yeah, these output examples looks good, thanks.